src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java

```diff
@@ -481,10 +497,36 @@ protected SCMRevision retrieve(SCMHead head, TaskListener listener) throws IOExc
 
     @Override
     public SCM build(SCMHead head, SCMRevision revision) {
+        BitbucketRepositoryType repositoryType;
+        if (head instanceof PullRequestSCMHead) {
+            repositoryType = ((PullRequestSCMHead) head).getRepositoryType();
+        } else if (head instanceof BranchSCMHead) {
+            repositoryType = ((BranchSCMHead) head).getRepositoryType();
+        } else {
+            throw new IllegalArgumentException("Either PullRequestSCMHead or BranchSCMHead required as parameter");
+        }
+        if (repositoryType == null) {
+            if (revision instanceof MercurialRevision) {
+                repositoryType = BitbucketRepositoryType.MERCURIAL;
+            } else if (revision instanceof AbstractGitSCMSource.SCMRevisionImpl) {
+                repositoryType = BitbucketRepositoryType.GIT;
+            } else {
+                try {
+                    repositoryType = getRepositoryType();
```

Comment:
Is the code from line 500 to here to avoid a potential HTTP request? If so, I think this is making the code unnecessarily complex.